pkgs/development/python-modules/codepy/default.nix

```diff
@@ -0,0 +1,27 @@
+{ lib
+, buildPythonPackage
+, fetchPypi
+, pytools
+, numpy
+, appdirs
+, six
+, cgen
+}:
+
+buildPythonPackage rec {
+  pname = "codepy";
+  version = "2019.1";
+
+  src = fetchPypi {
+    inherit pname version;
+    sha256 = "384f22c37fe987c0ca71951690c3c2fd14dacdeddbeb0fde4fd01cd84859c94e";
+  };
+
+  buildInputs = [ pytools numpy appdirs six cgen ];
+
```

Comment:
Missing tests, pythonImportsCheck and maintainer.

Reply:
The test fails, seems like it tries to create directories. Added everything either way.
```
============================= test session starts ==============================
platform linux -- Python 3.9.10, pytest-6.2.5, py-1.11.0, pluggy-1.0.0
rootdir: /build/codepy-2019.1
collected 2 items                                                              

test/test_identical_symbols.py x                                         [ 50%]
test/test_two_stage_compile.py F                                         [100%]

=================================== FAILURES ===================================
_____________________________________ test _____________________________________

    def test():
        toolchain = guess_toolchain()
    
        module_code = """
        extern "C" {
            int const greet()
            {
                return 1;
            }
        }
        """
        # compile to object file
>       _, _, obj_path, _ = compile_from_string(toolchain, 'module', module_code,
                                                object=True)

test/test_two_stage_compile.py:18: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 
/nix/store/36scsgiy8bf09b4g0bnknnmvihgwrazs-python3.9-codepy-2019.1/lib/python3.9/site-packages/codepy/jit.py:286: in compile_from_string
    os.makedirs(cache_dir)
/nix/store/hb1lzaisgx2m9n29hqhh6yp6hasplq1v-python3-3.9.10/lib/python3.9/os.py:215: in makedirs
    makedirs(head, exist_ok=exist_ok)
/nix/store/hb1lzaisgx2m9n29hqhh6yp6hasplq1v-python3-3.9.10/lib/python3.9/os.py:215: in makedirs
    makedirs(head, exist_ok=exist_ok)
/nix/store/hb1lzaisgx2m9n29hqhh6yp6hasplq1v-python3-3.9.10/lib/python3.9/os.py:215: in makedirs
    makedirs(head, exist_ok=exist_ok)
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 

name = '/homeless-shelter', mode = 511, exist_ok = False

    def makedirs(name, mode=0o777, exist_ok=False):
        """makedirs(name [, mode=0o777][, exist_ok=False])
    
        Super-mkdir; create a leaf directory and all intermediate ones.  Works like
        mkdir, except that any intermediate path segment (not just the rightmost)
        will be created if it does not exist. If the target directory already
        exists, raise an OSError if exist_ok is False. Otherwise no exception is
        raised.  This is recursive.
    
        """
        head, tail = path.split(name)
        if not tail:
            head, tail = path.split(head)
        if head and tail and not path.exists(head):
            try:
                makedirs(head, exist_ok=exist_ok)
            except FileExistsError:
                # Defeats race condition when another thread created the path
                pass
            cdir = curdir
            if isinstance(tail, bytes):
                cdir = bytes(curdir, 'ASCII')
            if tail == cdir:           # xxx/newdir/. exists if xxx/newdir exists
                return
        try:
>           mkdir(name, mode)
E           PermissionError: [Errno 13] Permission denied: '/homeless-shelter'

/nix/store/hb1lzaisgx2m9n29hqhh6yp6hasplq1v-python3-3.9.10/lib/python3.9/os.py:225: PermissionError
=============================== warnings summary ===============================
test/test_identical_symbols.py::test_identical_symbols
  /nix/store/36scsgiy8bf09b4g0bnknnmvihgwrazs-python3.9-codepy-2019.1/lib/python3.9/site-packages/codepy/jit.py:271: DeprecationWarning: wait_on_error is deprecated and has no effect
    warn("wait_on_error is deprecated and has no effect",

-- Docs: https://docs.pytest.org/en/stable/warnings.html
=========================== short test summary info ============================
FAILED test/test_two_stage_compile.py::test - PermissionError: [Errno 13] Per...
=================== 1 failed, 1 xfailed, 1 warning in 0.46s ====================
error: builder for '/nix/store/ym312b4w3p2hvxjvkggr60d8a83fc6wb-python3.9-codepy-2019.1.drv' failed with exit code 1;
       last 10 log lines:
       > /nix/store/hb1lzaisgx2m9n29hqhh6yp6hasplq1v-python3-3.9.10/lib/python3.9/os.py:225: PermissionError
       > =============================== warnings summary ===============================
       > test/test_identical_symbols.py::test_identical_symbols
       >   /nix/store/36scsgiy8bf09b4g0bnknnmvihgwrazs-python3.9-codepy-2019.1/lib/python3.9/site-packages/codepy/jit.py:271: DeprecationWarning: wait_on_error is deprecated and has no effect
       >     warn("wait_on_error is deprecated and has no effect",
       >
       > -- Docs: https://docs.pytest.org/en/stable/warnings.html
       > =========================== short test summary info ============================
       > FAILED test/test_two_stage_compile.py::test - PermissionError: [Errno 13] Per...
       > =================== 1 failed, 1 xfailed, 1 warning in 0.46s ====================
       For full logs, run 'nix log /nix/store/ym312b4w3p2hvxjvkggr60d8a83fc6wb-python3.9-codepy-2019.1.drv'.
     ```

Reply:
I think we want to do

```nix
preCheck = ''
  export HOME=$(mktemp -d)
'';
```

Reply:
```
============================= test session starts ==============================
platform linux -- Python 3.9.11, pytest-7.0.1, pluggy-1.0.0
rootdir: /build/source
collected 2 items                                                              

test/test_identical_symbols.py x                                         [ 50%]
test/test_two_stage_compile.py F                                         [100%]

=================================== FAILURES ===================================
_____________________________________ test _____________________________________

    def test():
        toolchain = guess_toolchain()
    
        module_code = """
        extern "C" {
            int const greet()
            {
                return 1;
            }
        }
        """
        # compile to object file
>       _, _, obj_path, _ = compile_from_string(toolchain, 'module', module_code,
                                                object=True)

test/test_two_stage_compile.py:18: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 
/nix/store/kqb6h8hf3nag7nana94j5aw0b2aq81rh-python3.9-codepy-2019.1/lib/python3.9/site-packages/codepy/jit.py:286: in compile_from_string
    os.makedirs(cache_dir)
/nix/store/9px00aaqzb6n5p03i9wd8rx3msg95y9r-python3-3.9.11/lib/python3.9/os.py:215: in makedirs
    makedirs(head, exist_ok=exist_ok)
/nix/store/9px00aaqzb6n5p03i9wd8rx3msg95y9r-python3-3.9.11/lib/python3.9/os.py:215: in makedirs
    makedirs(head, exist_ok=exist_ok)
/nix/store/9px00aaqzb6n5p03i9wd8rx3msg95y9r-python3-3.9.11/lib/python3.9/os.py:215: in makedirs
    makedirs(head, exist_ok=exist_ok)
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 

name = '/homeless-shelter', mode = 511, exist_ok = False

    def makedirs(name, mode=0o777, exist_ok=False):
        """makedirs(name [, mode=0o777][, exist_ok=False])
    
        Super-mkdir; create a leaf directory and all intermediate ones.  Works like
        mkdir, except that any intermediate path segment (not just the rightmost)
        will be created if it does not exist. If the target directory already
        exists, raise an OSError if exist_ok is False. Otherwise no exception is
        raised.  This is recursive.
    
        """
        head, tail = path.split(name)
        if not tail:
            head, tail = path.split(head)
        if head and tail and not path.exists(head):
            try:
                makedirs(head, exist_ok=exist_ok)
            except FileExistsError:
                # Defeats race condition when another thread created the path
                pass
            cdir = curdir
            if isinstance(tail, bytes):
                cdir = bytes(curdir, 'ASCII')
            if tail == cdir:           # xxx/newdir/. exists if xxx/newdir exists
                return
        try:
>           mkdir(name, mode)
E           PermissionError: [Errno 13] Permission denied: '/homeless-shelter'

/nix/store/9px00aaqzb6n5p03i9wd8rx3msg95y9r-python3-3.9.11/lib/python3.9/os.py:225: PermissionError
=============================== warnings summary ===============================
test/test_identical_symbols.py::test_identical_symbols
  /nix/store/kqb6h8hf3nag7nana94j5aw0b2aq81rh-python3.9-codepy-2019.1/lib/python3.9/site-packages/codepy/jit.py:271: DeprecationWarning: wait_on_error is deprecated and has no effect
    warn("wait_on_error is deprecated and has no effect",

-- Docs: https://docs.pytest.org/en/stable/how-to/capture-warnings.html
=========================== short test summary info ============================
FAILED test/test_two_stage_compile.py::test - PermissionError: [Errno 13] Per...
=================== 1 failed, 1 xfailed, 1 warning in 0.45s ====================
error: builder for '/nix/store/syx0max6a9jgpxkv1c0mmvvl77l905hq-python3.9-codepy-2019.1.drv' failed with exit code 1;
       last 10 log lines:
       > /nix/store/9px00aaqzb6n5p03i9wd8rx3msg95y9r-python3-3.9.11/lib/python3.9/os.py:225: PermissionError
       > =============================== warnings summary ===============================
       > test/test_identical_symbols.py::test_identical_symbols
       >   /nix/store/kqb6h8hf3nag7nana94j5aw0b2aq81rh-python3.9-codepy-2019.1/lib/python3.9/site-packages/codepy/jit.py:271: DeprecationWarning: wait_on_error is deprecated and has no effect
       >     warn("wait_on_error is deprecated and has no effect",
       >
       > -- Docs: https://docs.pytest.org/en/stable/how-to/capture-warnings.html
       > =========================== short test summary info ============================
       > FAILED test/test_two_stage_compile.py::test - PermissionError: [Errno 13] Per...
       > =================== 1 failed, 1 xfailed, 1 warning in 0.45s ====================
       For full logs, run 'nix log /nix/store/syx0max6a9jgpxkv1c0mmvvl77l905hq-python3.9-codepy-2019.1.drv'.
       ```
       Same error :/

Reply:
Hmmm, then please remove the checkInputs related code again and add a ``doCheck = false;`` with a comment that the tests are broken.